### PR TITLE
feat: add in flight connection middleware

### PR DESCRIPTION
**Pull request by tomMoulard**
### What does this PR do?

This PR introduces a new TCP Middleware called InFlightConn.

This middleware does the same as the HTTP counterpart (i.e. `amount`), except that you can't have `sourceCriterion` since you have less metadata about a connection in TCP than in an HTTP request. Furthermore, the TCP connections are grouped by remote IP.

### Motivation

Solve half of #5999

### More

- [X] Added/updated tests
- [X] Added/updated documentation

### CODE REVIEW

File: docs/content/middlewares/tcp/overview.md [hunk 1]
```diff
@@ -131,4 +131,5 @@ tcp:
 
 | Middleware                                | Purpose                                           | Area                        |
 |-------------------------------------------|---------------------------------------------------|-----------------------------|
-| [IPWhiteList](ipwhitelist.md)             | Limit the allowed client IPs                      | Security, Request lifecycle |
+| [InFilightConn](inflightconn.md)          | Limits the number of simultaneous connections.    | Security, Request lifecycle |
```
Comment (mpl) [inline docs/content/middlewares/tcp/overview.md:134]:
```suggestion
| [InFlightConn](inflightconn.md)          | Limits the number of simultaneous connections.    | Security, Request lifecycle |
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	name           string
+	next           tcp.Handler
+	maxConnections int64
+
+	mu          sync.Mutex
+	connections map[string]int64 // current number of connections by remote IP.
+}
+
+// New creates a max connections middleware.
+// The connections are limited by remote IP.
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:27]:
```suggestion
// The connections are identified and grouped by remote IP.
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+	i.next.ServeTCP(conn)
+}
+
+func (i *inFlightConn) acquire(ip string) error {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	if i.connections[ip] >= i.maxConnections {
+		return fmt.Errorf("max connection reached for %s", ip)
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:69]:
```suggestion
		return fmt.Errorf("max number of connections reached for %s", ip)
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		conn.Close()
+		return
+	}
+
+	defer i.release(ip)
+
+	i.next.ServeTCP(conn)
+}
+
+func (i *inFlightConn) acquire(ip string) error {
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:64]:
```suggestion
func (i *inFlightConn) increment(ip string) error {
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	if i.connections[ip] >= i.maxConnections {
+		return fmt.Errorf("max connection reached for %s", ip)
+	}
+
+	i.connections[ip]++
+
+	return nil
+}
+
+func (i *inFlightConn) release(ip string) {
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:77]:
```suggestion
func (i *inFlightConn) decrement(ip string) {
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	i.connections[ip]--
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:81]: Just to be safe, we might want to guarantee that if release is called more often than acquire, we do not let this number go below 0.

File: pkg/middlewares/tcp/inflightconn/inflight_conn_test.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	})
+
+	middleware, err := New(context.Background(), next, dynamic.TCPInFlightConn{Amount: 1}, "foo")
+	require.NoError(t, err)
+
+	// The first connection should succeed and wait.
+	go middleware.ServeTCP(fakeConn{addr: "127.0.0.1:9000", wait: true})
+	requireMessage(t, proceedCh)
+
+	// The second connection should be closed as the maximum number of connections is exceeded.
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn_test.go:37]:
```suggestion
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn_test.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	closeCh := make(chan struct{})
+
+	go middleware.ServeTCP(fakeConn{addr: "127.0.0.1:9001", closeCh: closeCh})
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn_test.go:40]:
```suggestion
	// The second connection from the same remote address should be closed as the maximum number of connections is exceeded.
	go middleware.ServeTCP(fakeConn{addr: "127.0.0.1:9000", closeCh: closeCh})
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn_test.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	requireMessage(t, closeCh)
+
+	// The connection with another remote address should succeed.
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn_test.go:43]:
```suggestion
	// The connection from another remote address should succeed.
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn_test.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	go middleware.ServeTCP(fakeConn{addr: "127.0.0.2:9000"})
+	requireMessage(t, proceedCh)
+
+	// Once the first connection is closed, next connection with the same remote address should succeed.
+	close(waitCh)
+	requireMessage(t, finishCh)
+
+	go middleware.ServeTCP(fakeConn{addr: "127.0.0.1:9002"})
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn_test.go:51]:
```suggestion
	go middleware.ServeTCP(fakeConn{addr: "127.0.0.1:9000"})
```

**Review by mpl — CHANGES_REQUESTED** [review]

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+		return fmt.Errorf("max number of connections reached for %s", ip)
+	}
+
+	i.connections[ip]++
+
+	return nil
+}
+
+// decrement allow for a new connection for a given IP.
+// It ensures that the number of connection is greater or equal to zero.
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:80]:
```suggestion
// decrement decreases the counter for the number of connections tracked for the
// given IP.
// It ensures that the counter does not go below zero.
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+		return
+	}
+
+	defer i.decrement(ip)
+
+	i.next.ServeTCP(conn)
+}
+
+// increment check if an IP can open a new connection.
+// Returns an error if it cannot.
```
Comment (mpl) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:65]:
```suggestion
// increment increases the counter for the number of connections tracked for the
// given IP.
// It returns an error if the counter would go above the max allowed number of
// connections.
```

**Review by mpl — CHANGES_REQUESTED** [review]

File: docs/content/middlewares/tcp/overview.md [hunk 13]
```diff
@@ -131,4 +131,5 @@ tcp:
 
 | Middleware                                | Purpose                                           | Area                        |
 |-------------------------------------------|---------------------------------------------------|-----------------------------|
-| [IPWhiteList](ipwhitelist.md)             | Limit the allowed client IPs                      | Security, Request lifecycle |
+| [InFlightConn](inflightconn.md)          | Limits the number of simultaneous connections.    | Security, Request lifecycle |
+| [IPWhiteList](ipwhitelist.md)             | Limit the allowed client IPs.                     | Security, Request lifecycle |
```
Comment (kevinpollet) [inline docs/content/middlewares/tcp/overview.md:135]:
```suggestion
| [InFlightConn](inflightconn.md)           | Limits the number of simultaneous connections.    | Security, Request lifecycle |
| [IPWhiteList](ipwhitelist.md)             | Limit the allowed client IPs.                     | Security, Request lifecycle |
```

File: pkg/middlewares/tcp/inflightconn/inflight_conn.go [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+	ip, _, err := net.SplitHostPort(conn.RemoteAddr().String())
+	if err != nil {
+		logger.Errorf("Cannot parse IP from remote addr: %v", err)
+		conn.Close()
+		return
+	}
+
+	err = i.increment(ip)
+	if err != nil {
```
Comment (kevinpollet) [inline pkg/middlewares/tcp/inflightconn/inflight_conn.go:53]:
```suggestion
	if err = i.increment(ip); err != nil {
```

**Review by mpl — APPROVED** [review]

**Review by rtribotte — APPROVED** [review]
Thanks 👍

**Pull request merged**